contrib/storage-hbase/src/main/java/org/apache/drill/exec/store/hbase/HBaseConnectionManager.java

```diff
@@ -0,0 +1,92 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.drill.exec.store.hbase;
+
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.drill.common.exceptions.UserException;
+import org.apache.drill.exec.store.hbase.HBaseStoragePlugin.HBaseConnectionKey;
+import org.apache.hadoop.hbase.client.Connection;
+import org.apache.hadoop.hbase.client.ConnectionFactory;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.cache.RemovalListener;
+import com.google.common.cache.RemovalNotification;
+import com.google.common.util.concurrent.UncheckedExecutionException;
+
+/**
+ * <p>A singleton class which manages the lifecycle of HBase connections.</p>
+ * <p>One connection per storage plugin instance is maintained.</p>
+ */
+public final class HBaseConnectionManager
+    extends CacheLoader<HBaseConnectionKey, Connection> implements RemovalListener<HBaseConnectionKey, Connection> {
+  private static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(HBaseConnectionManager.class);
+
+  public static final HBaseConnectionManager INSTANCE = new HBaseConnectionManager();
+
+  private final LoadingCache<HBaseConnectionKey, Connection> connectionCache;
+
+  private HBaseConnectionManager() {
+    this.connectionCache = CacheBuilder.newBuilder()
+        .expireAfterAccess(24, TimeUnit.HOURS)
+        .removalListener(this)
+        .build(this);
+  }
+
+  @Override
+  public Connection load(HBaseConnectionKey key) throws Exception {
+    Connection connection = ConnectionFactory.createConnection(key.getHBaseConf());
+    logger.debug("HBase connection '{}' created.", connection);
+    return connection;
+  }
+
+  @Override
+  public void onRemoval(RemovalNotification<HBaseConnectionKey, Connection> notification) {
+    try {
+      Connection conn = notification.getValue();
+      if (conn != null) {
+        conn.close();
+      }
+      logger.debug("HBase connection '{}' closed.", notification.getValue());
+    } catch (Throwable t) {
+      logger.warn("Error while closing HBase connection.", t);
+    }
+  }
+
+  public Connection getConnection(HBaseConnectionKey key) {
+    try {
+      Connection conn = connectionCache.get(key);
+      if (conn.isAborted()
+          || conn.isClosed()) {
+        connectionCache.invalidate(key);
```

Comment:
Thanks for catching this, not sure how it slipped through my own review :(

Couldn't find a simple way to make Guava do this (validate a cached value before returning), so will change it to do both eviction and load atomically in this code itself.